src/core/structure/sheet.ts

```diff
@@ -396,18 +439,107 @@ export default class Sheet {
     return this.cellData.get(cellKey)!;
   }
 
-  private resolveCell(cell: Cell): boolean {
-    const value = this.expressionHandler.evaluate(cell.formula);
-    if (value == null) {
-      cell.state = CellState.INVALID_EXPRESSION;
-      return false;
+  private resolveCell(
+    cell: Cell,
+    colKey: ColumnKey,
+    rowKey: RowKey,
+  ): ResolveCellResult {
+    const evalResult = this.expressionHandler.evaluate(cell.formula);
+    if (!evalResult) {
+      const prevState = cell.state;
+      cell.setState(CellState.INVALID_EXPRESSION);
+      return { valueChanged: prevState == CellState.OK }; // The cell's value has changed if the state changed from OK to invalid.
     }
 
     // TODO Resolve restrictions from cell formatting here (CellState.INVALID_FORMAT).
 
-    cell.state = CellState.OK;
-    cell.value = value;
-    return cell.formula != cell.value;
+    cell.setState(CellState.OK);
+
+    const valueChanged = cell.value != evalResult.value;
+    cell.value = evalResult.value;
+
+    // Update referencesOut of this cell and referencesIn of newly referenced cells.
+    const oldOut = new Map<CellKey, PositionInfo>(cell.referencesOut);
+    cell.referencesOut.clear();
+    evalResult.references.forEach((ref) => {
+      // Add referred cells to this cell's referencesOut.
+      const referredCell = this.getCell(ref.columnKey!, ref.rowKey!)!;
+      cell.referencesOut.set(referredCell.key, {
+        columnKey: ref.columnKey!,
+        rowKey: ref.rowKey!,
+      });
+
+      // Add this cell to the referred cell's referencesIn.
+      referredCell.referencesIn.set(cell.key, {
+        columnKey: colKey,
+        rowKey: rowKey,
+      });
+    });
+
+    // Resolve (oldOut - newOut) to get references that were removed from the formula.
+    const removedReferences = new Map<CellKey, PositionInfo>(
+      [...oldOut].filter(([cellKey]) => !cell.referencesOut.has(cellKey)),
+    );
+
+    // Clean up the referencesIn sets of cells that are no longer referenced by the formula.
+    removedReferences.forEach((_, cellKey) => {
+      const referredCell = this.cellData.get(cellKey)!;
+      referredCell.referencesIn.delete(cell.key);
+    });
+
+    // After references are updated, check for circular references.
+    if (evalResult.references.length && this.hasCircularReference(cell)) {
+      cell.setState(CellState.CIRCULAR_REFERENCE);
+    }
+
+    // If the value of the cell hasn't changed, there's no need to update cells that reference this cell.
+    if (!valueChanged && cell.state == CellState.OK)
+      return { valueChanged: false };
+
+    // Return the set of cells whose values has changed as a result of updating this cell.
+    const dirtyCells = new Map<CellKey, PositionInfo>();
```

Comment:
We are not setting any value to this map. Is there something missing here?

Reply:
This was left over from a previous approach; `ResolveCellResult` is unused currently.